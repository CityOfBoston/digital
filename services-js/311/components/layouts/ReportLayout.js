// @flow
/* eslint react/no-unused-prop-types: 0 */

import React from 'react';

import Nav from '../common/Nav';
import LocationMap from '../location';
import StartFormContainer from '../report-start';
import ReportFormContainer from '../report-form';

import type { LoopbackGraphql } from '../../data/graphql/loopback-graphql';
import type { Service, ServiceSummary } from '../../data/types';

type ServiceProps = {
  view: 'service',
  code: string,
  service: ?Service,
  pickLocation: boolean,
};

type SummaryProps = {
  view: 'summaries',
  serviceSummaries: ?ServiceSummary[],
  pickLocation: boolean,
};

export type InitialProps = ServiceProps | SummaryProps;

export type Props = {
  initialProps: InitialProps,
  loopbackGraphql: LoopbackGraphql,
  showServiceForm: (ServiceSummary) => void,
  goToReportForm: () => void,
}

export default class ReportLayout extends React.Component {
  props: Props;

  static getTitle(initialProps: InitialProps) {
    switch (initialProps.view) {
      case 'summaries': return 'Report a Problem';
      case 'service': {
        const { service, pickLocation } = initialProps;

        if (pickLocation) {
          return 'Choose location';
        } else if (service) {
          return service.name;
        } else {
          return 'Not found';
        }
      }
      default: return '';
    }
  }

  render() {
    const { initialProps: { pickLocation }, goToReportForm } = this.props;

    return (
      <div>
        <Nav />
        <LocationMap active={pickLocation} goToReportForm={goToReportForm} />

        {this.renderContent()}
      </div>
    );
  }

  renderContent() {
    const { initialProps, showServiceForm, loopbackGraphql } = this.props;

    switch (initialProps.view) {
      case 'summaries':
        return (
          <StartFormContainer serviceSummaries={initialProps.serviceSummaries} showServiceForm={showServiceForm} />
        );

      case 'service': {
        const { service, pickLocation } = initialProps;
        if (pickLocation) {
          return null;
        } else {
          return <ReportFormContainer service={service} loopbackGraphql={loopbackGraphql} />;
        }
      }

      default:
        return null;
    }
  }
}
