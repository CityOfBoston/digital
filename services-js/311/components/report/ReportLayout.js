// @flow
/* global liveagent */
/* eslint jsx-a11y/anchor-has-content: 0 */

import React from 'react';
import { css } from 'glamor';
import Router from 'next/router';
import type { Context } from 'next';
import { action, computed } from 'mobx';
import { observer } from 'mobx-react';

import type { RequestAdditions } from '../../server/next-handlers';

import Nav from '../common/Nav';
import LocationMap from './map/LocationMap';
import type { MapMode } from './map/LocationMap';
import HomeDialog from './home/HomeDialog';
import RequestDialog from './request/RequestDialog';

import { MEDIA_LARGE, HEADER_HEIGHT } from '../style-constants';

import makeLoopbackGraphql from '../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../data/dao/loopback-graphql';

import loadTopServiceSummaries from '../../data/dao/load-top-service-summaries';
import loadService from '../../data/dao/load-service';

import type { Service, ServiceSummary } from '../../data/types';

import getStore from '../../data/store';
import type { AppStore } from '../../data/store';

type HomeData = {|
  view: 'home',
  topServiceSummaries: ServiceSummary[],
  stage: 'home' | 'choose',
|};

type RequestData = {|
  view: 'request',
  code: string,
  service: ?Service,
  stage: 'questions' | 'location' | 'contact' | 'submit',
|};

export type InitialProps = {|
  data: HomeData | RequestData,
|};

export type Props = {|
  store: AppStore,
  /* :: ...InitialProps, */
|};

// This is the main container for content. We want to be at least full height on
// large screens to push the footer down to where you need to scroll for it.
const CONTAINER_STYLE = css({
  minHeight: 0,
  position: 'relative',
  justifyContent: 'flex-start',
  [MEDIA_LARGE]: {
    minHeight: '100vh',
  },
});

const BACKGROUND_MAP_CONTAINER_STYLE = css({
  position: 'fixed',
  width: '100%',
  top: HEADER_HEIGHT,
  bottom: 0,
  background: '#9B9B9B',
});

// We have one class for picking the service type and doing the entire request
// so that we can keep a consistent Map background behind the
// dialog and not repeatedly re-mount it.
//
// This class has two "views": the "summaries" view that lets the user pick a
// service request, and the "service" view for the request flow. That flow
// proceeds through the "questions" "location" and "contact" stages.
@observer
export default class ReportLayout extends React.Component {
  props: Props;
  state: {
    locationMapActive: boolean,
  }
  loopbackGraphql: LoopbackGraphql;

  static async getInitialProps({ query, req, res }: Context<RequestAdditions>): Promise<InitialProps> {
    const loopbackGraphql = makeLoopbackGraphql(req);

    let data;

    if (query.code) {
      data = await ReportLayout.getRequestData(query, res, getStore(), loopbackGraphql);
    } else {
      data = await ReportLayout.getHomeData(query, loopbackGraphql);
    }

    return {
      data,
    };
  }

  static async getHomeData({ stage }, loopbackGraphql): Promise<HomeData> {
    const store = getStore();

    return {
      view: 'home',
      topServiceSummaries: store.topServiceSummaries || await loadTopServiceSummaries(loopbackGraphql),
      stage: stage === 'choose' ? stage : 'home',
    };
  }

  static async getRequestData({ code, stage }, res, store, loopbackGraphql): Promise<RequestData> {
    let service = store.serviceCache.get(code);

    if (!service) {
      service = await loadService(loopbackGraphql, code);

      if (!service && res) {
        res.statusCode = 404;
      }
    }

    stage = stage || 'questions';

    switch (stage) {
      case 'questions':
      case 'location':
      case 'contact':
      case 'submit':
        return {
          view: 'request',
          code,
          service,
          stage,
        };
      default:
        throw new Error(`Unknown stage: ${stage}`);
    }
  }

  // TODO(finneganh): Move service cache and lookup out of this class
  @action static addServiceToCache(store: AppStore, service: Service) {
    store.serviceCache.set(service.code, service);
  }

  constructor(props: Props) {
    super(props);

    this.loopbackGraphql = makeLoopbackGraphql();

    this.updateStoreWithProps(props);

    this.state = {
      locationMapActive: false,
    };
  }

  componentWillReceiveProps(props: Props) {
    this.updateStoreWithProps(props);
    this.setState({ locationMapActive: false });
  }

  startChat = () => {
    const { liveAgentButtonId } = this.props.store;
    liveagent.startChat(liveAgentButtonId);
  }

  @computed get mapActivationRatio(): number {
    const { store: { ui }, data: { view, stage } } = this.props;
    if (view === 'home' && stage === 'home') {
      return Math.min(1.0, ui.scrollY / (ui.visibleHeight * 0.75));
    } else {
      return 0;
    }
  }

  @action
  updateStoreWithProps(props: Props) {
    const { store, data } = props;

    switch (data.view) {
      case 'home':
        store.topServiceSummaries = data.topServiceSummaries;
        store.currentService = null;
        break;

      case 'request': {
        const { service } = data;
        if (service) {
          ReportLayout.addServiceToCache(store, service);
        }
        store.currentService = service;
        break;
      }

      default:
        break;
    }
  }

  routeToServiceForm = async (code: string, stage: string = 'questions') => {
    if (stage === 'questions') {
      await Router.push(`/report?code=${code}`, `/report/${code}`);
    } else {
      await Router.push(`/report?code=${code}&stage=${stage}`, `/report/${code}/${stage}`);
    }

    window.scrollTo(0, 0);
  }

  setLocationMapActive = (active: boolean) => {
    const { locationMapActive } = this.state;

    if (locationMapActive !== active) {
      this.setState({ locationMapActive: active });
    }
  }

  render() {
    const { data, store } = this.props;
    const { locationMapActive } = this.state;
    const { ui: { mediaLarge } } = store;

    let mapMode: MapMode;
    if (locationMapActive) {
      mapMode = 'picker';
    } else if (this.mapActivationRatio === 1.0) {
      mapMode = 'requests';
    } else {
      mapMode = 'inactive';
    }

    return (
      <div>
        <Nav activeSection="report" />

        {/* Outer box needs to take up at least the screen size on desktop, so
            that the content can center over the map and keep the footer from
            encroaching up. */}
        <div className={`mn mn--full mn--nv-s ${CONTAINER_STYLE.toString()}`} style={{ backgroundColor: 'transparent' }}>
          { mediaLarge &&
            <div className={BACKGROUND_MAP_CONTAINER_STYLE}>
              <LocationMap
                store={store}
                loopbackGraphql={this.loopbackGraphql}
                mode={mapMode}
                opacityRatio={this.mapActivationRatio}
              />
            </div>
          }

          { /* We can't put any wrappers around the dialogs below because in the
               case of RequestDialog's location picker, it needs to make sure nothing
               is overlapping the map so it can get clicks. */}

          { data.view === 'home' &&
            <HomeDialog
              store={store}
              loopbackGraphql={this.loopbackGraphql}
              stage={data.stage}
            /> }

          { data.view === 'request' &&
            <RequestDialog
              store={store}
              stage={data.stage}
              loopbackGraphql={this.loopbackGraphql}
              routeToServiceForm={this.routeToServiceForm}
              setLocationMapActive={this.setLocationMapActive}
            />}
        </div>

      </div>
    );
  }
}
