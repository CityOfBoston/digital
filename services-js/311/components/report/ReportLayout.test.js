// @flow
/* eslint no-fallthrough: 0 */

import React from 'react';
import renderer from 'react-test-renderer';
import Router from 'next/router';

import type { Service, ServiceSummary } from '../../data/types';

import { makeServerContext } from '../../lib/test/make-context';

import getStore from '../../data/store';

import ReportLayout from './ReportLayout';

jest.mock('next/router');

jest.mock('../../data/dao/load-top-service-summaries');
jest.mock('../../data/dao/load-service');
jest.mock('../../data/dao/search-requests');

const loadTopServiceSummaries: JestMockFn = (require('../../data/dao/load-top-service-summaries'): any).default;
const loadService: JestMockFn = (require('../../data/dao/load-service'): any).default;
const searchRequests: JestMockFn = (require('../../data/dao/search-requests'): any).default;

const MOCK_SERVICE_SUMMARIES: ServiceSummary[] = [{
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  description: 'Something is threatening the fabric of the universe',
  group: 'ultimate',
}];

const MOCK_SERVICE: Service = {
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  contactRequirement: 'REQUIRED',
  locationRequirement: 'VISIBLE',
  attributes: [{
    required: false,
    type: 'TEXT',
    code: 'ST-CMTS',
    description: 'Please provide any other relevant information:',
    values: null,
    conditionalValues: null,
    dependencies: null,
  }, {
    required: false,
    type: 'STRING',
    code: 'INFO-CSIRMV1',
    description: '**All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**',
    values: null,
    conditionalValues: null,
    dependencies: null,
  }, {
    required: true,
    type: 'SINGLEVALUELIST',
    code: 'SR-CSIRMV1',
    description: 'How many dimensions have been breached?',
    values: [{ key: 'One', name: 'One' }, { key: 'Two', name: 'Two' }, { key: 'Three', name: 'Three' }, { key: 'More than Three', name: 'More than Three' }],
    conditionalValues: [],
    dependencies: null,
  }],
};

let store;

beforeEach(() => {
  store = getStore();
  store.liveAgentAvailable = true;
  store.ui.visibleWidth = 1300;
  searchRequests.mockReturnValue(new Promise(() => {}));
});

describe('report form', () => {
  let data;

  beforeEach(async () => {
    loadTopServiceSummaries.mockReturnValue(MOCK_SERVICE_SUMMARIES);

    const ctx = makeServerContext('/report');
    data = (await ReportLayout.getInitialProps(ctx)).data;
  });

  test('getInitialProps', () => {
    expect(data.view).toEqual('home');
  });

  test('rendering', () => {
    const component = renderer.create(
      <ReportLayout store={store} data={data} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('existing service page', () => {
  beforeEach(() => {
    loadService.mockReturnValue(MOCK_SERVICE);
  });

  test('getInitialProps', async () => {
    const ctx = makeServerContext('/report', { code: 'CSMCINC' });
    const data = (await ReportLayout.getInitialProps(ctx)).data;
    expect(data.view).toEqual('request');
  });

  test('rendering', async () => {
    const ctx = makeServerContext('/report', { code: 'CSMCINC' });
    const data = (await ReportLayout.getInitialProps(ctx)).data;
    const component = renderer.create(
      <ReportLayout data={data} store={store} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('rendering phone', async () => {
    const ctx = makeServerContext('/report', { code: 'CSMCINC' }, { isPhone: true });
    const data = (await ReportLayout.getInitialProps(ctx)).data;
    const component = renderer.create(
      <ReportLayout data={data} store={store} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('rendering phone location step', async () => {
    const ctx = makeServerContext('/report', { code: 'CSMCINC', stage: 'location' }, { isPhone: true });
    const data = (await ReportLayout.getInitialProps(ctx)).data;
    const component = renderer.create(
      <ReportLayout data={data} store={store} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('missing service page', () => {
  let ctx;
  let data;

  beforeEach(async () => {
    loadService.mockReturnValue(null);

    ctx = makeServerContext('/report', { code: 'CSMCINC' });
    data = (await ReportLayout.getInitialProps(ctx)).data;
  });

  test('getInitialProps', () => {
    // flow check
    if (ctx.res) {
      expect(ctx.res.statusCode).toEqual(404);
    } else {
      expect(ctx.res).toBeDefined();
    }

    switch (data.view) {
      case 'request':
        expect(data.props.service).toBeNull();
      default:
        expect(data.view).toEqual('request');
    }
  });

  test('rendering', () => {
    const component = renderer.create(
      <ReportLayout data={data} store={store} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('routeToServiceForm', () => {
  let reportLayout;

  beforeEach(() => {
    const props: any = {
      data: {},
      store,
    };
    reportLayout = new ReportLayout(props);
  });

  it('defaults to questions', () => {
    reportLayout.routeToServiceForm('CSMCINC');
    expect(Router.push).toHaveBeenCalledWith('/report?code=CSMCINC', '/report/CSMCINC');
  });

  it('includes the stage', () => {
    reportLayout.routeToServiceForm('CSMCINC', 'contact');
    expect(Router.push).toHaveBeenCalledWith('/report?code=CSMCINC&stage=contact', '/report/CSMCINC/contact');
  });
});
