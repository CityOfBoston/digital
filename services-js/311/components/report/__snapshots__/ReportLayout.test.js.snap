// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`existing service page rendering 1`] = `
<div
  className={
    Object {
      "data-css-awku5i": "",
    }
  }
>
  <div
    className="nav ft-ll css-gf15pf"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}
    >
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}
    >
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}
    >
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-1nvfmj1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1ldikjo": "",
        }
      }
    >
      <div
        className={
          Object {
            "data-css-1f2iwx6": "",
          }
        }
      >
        <div>
          <div
            className={
              Object {
                "data-css-dgukzt": "",
              }
            }
          >
            
          </div>
          <label>
            <p>
              Please provide any other relevant information:
               
            </p>
            <textarea
              className={
                Object {
                  "data-css-1bm19ha": "",
                }
              }
              name="ST-CMTS"
              onChange={[Function]}
              value=""
            />
          </label>
          <label>
            <p>
              **All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**
               
            </p>
            <input
              className={
                Object {
                  "data-css-rqg0fb": "",
                }
              }
              name="INFO-CSIRMV1"
              onChange={[Function]}
              type="text"
              value=""
            />
          </label>
          <label>
            <p>
              How many dimensions have been breached?
               
              (required)
            </p>
            <select
              name="SR-CSIRMV1"
              onChange={[Function]}
              value={null}
            >
              <option
                disabled={true}
                selected={true}
              >
                Please choose
              </option>
              <option
                disabled={true}
              >
                --------------------------
              </option>
              <option
                value="One"
              >
                One
              </option>
              <option
                value="Two"
              >
                Two
              </option>
              <option
                value="Three"
              >
                Three
              </option>
              <option
                value="More than Three"
              >
                More than Three
              </option>
            </select>
          </label>
          <button
            disabled={false}
            onClick={[Function]}
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`missing service page rendering 1`] = `
<div
  className={
    Object {
      "data-css-awku5i": "",
    }
  }
>
  <div
    className="nav ft-ll css-gf15pf"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}
    >
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}
    >
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}
    >
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-1nvfmj1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1ldikjo": "",
        }
      }
    >
      <div
        className={
          Object {
            "data-css-1f2iwx6": "",
          }
        }
      >
        <div
          className="sh sh--y"
          style={
            Object {
              "marginBottom": 40,
            }
          }
        >
          <h2
            className="sh-title"
          >
            Service not found
          </h2>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`report form rendering 1`] = `
<div
  className={
    Object {
      "data-css-awku5i": "",
    }
  }
>
  <div
    className="nav ft-ll css-gf15pf"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Report a problem
    </a>
    <a
      href="/lookup"
      onClick={[Function]}
    >
      Case look up
    </a>
    <a
      href="/go"
      onClick={[Function]}
    >
      311 on the go
    </a>
    <a
      href="/faq"
      onClick={[Function]}
    >
      FAQ
    </a>
  </div>
  <div
    className={
      Object {
        "data-css-1nvfmj1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1f2iwx6": "",
        }
      }
    >
      <div
        className="sh sh--y"
        style={
          Object {
            "marginBottom": 40,
          }
        }
      >
        <h2
          className="sh-title"
        >
          311: Boston City Services
        </h2>
      </div>
      <div
        className={
          Object {
            "data-css-9sglvw": "",
          }
        }
      >
        <textarea
          className={
            Object {
              "data-css-1j8x7ul": "",
            }
          }
          defaultValue=""
          name="description"
          onInput={[Function]}
          placeholder="How can we help?"
        />
        <ul
          className={
            Object {
              "data-css-18escyg": "",
            }
          }
        >
          <li
            className={
              Object {
                "data-css-14nloni": "",
              }
            }
          >
            <a
              className={
                Object {
                  "data-css-3br7c8": "",
                }
              }
              href="/report/CSMCINC/location"
              onClick={[Function]}
            >
              Cosmic Incursion
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
